### PR TITLE
Creating Project Module

### PR DESCRIPTION
**Pull request by mokot**
Creating Project Module and endpoints:
get all projects, get project by id or name, create project, update
project by id or name, delete project by id or name.

### CODE REVIEW

File: src/model/dto/project.create.dto.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+export class CreateProjectDto {
+
+  @ApiProperty({ example: "Reportko Project" })
+  @IsNotEmpty()
+  @IsString()
+  name: string;
+
+  @ApiProperty({ example: TeamKey.STUDENT })
+  @IsNotEmpty()
+  team: TeamKey;
```
Comment (norikkozmelj) [inline src/model/dto/project.create.dto.ts:24]: It is better to use team_id than team key. Try avoiding using keys.

File: src/model/dto/project.update.dto.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+export class UpdateProjectDto {
+
+  @ApiProperty({ example: "Reportko Project" })
+  @IsOptional()
+  @IsString()
+  name: string;
+
+  @ApiProperty({ example: TeamKey.STUDENT })
+  @IsOptional()
+  team: TeamKey;
```
Comment (norikkozmelj) [inline src/model/dto/project.update.dto.ts:24]: Same as on Post DTO

File: src/entity/project.entity.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  name: string;
+
+  @ApiProperty()
+  @ManyToOne(() => Team)
+  @JoinColumn({ name: 'team_id' })
+  @Type(() => Team)
+  @Transform((team) => team.value.key)
+  team: Team;
+
+  @Exclude()
```
Comment (norikkozmelj) [inline src/entity/project.entity.ts:46]: You can remove this, as you renamed team to team_id in line 41.

File: src/entity/project.entity.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+@Entity()
+export class Project {
+
+  @ApiProperty({ example: 1 })
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @ApiProperty({ example: "Reportko Project" })
+  @Column({ type: 'varchar', nullable: false, unique: true })
```
Comment (norikkozmelj) [inline src/entity/project.entity.ts:35]: nullable: false is by default, only use this attribute when you set it to true

File: src/entity/project.entity.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  @IsString()
+  name: string;
+
+  @ApiProperty()
+  @ManyToOne(() => Team)
+  @JoinColumn({ name: 'team_id' })
+  @Type(() => Team)
+  @Transform((team) => team.value.key)
```
Comment (norikkozmelj) [inline src/entity/project.entity.ts:43]: Remove transform as we will use whole object on frontend.

File: src/entity/project.entity.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  team: Team;
+
+  @Exclude()
+  @Column()
+  @IsNumber()
+  team_id: number;
+
+  @ApiProperty({ example: true })
+  @Column({ type: 'boolean', nullable: true, default: true })
```
Comment (norikkozmelj) [inline src/entity/project.entity.ts:52]: This cannot be nullable.

File: src/model/dto/project.create.dto.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  @ApiProperty({ example: true })
+  @IsOptional()
```
Comment (norikkozmelj) [inline src/model/dto/project.create.dto.ts:27]: this is mandatory, not optional.

File: src/model/dto/project.update.dto.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import { ApiProperty } from "@nestjs/swagger";
+import { IsBoolean, IsOptional, IsString } from "class-validator";
+import { TeamKey } from "src/enum/team-key.enum";
+
+export class UpdateProjectDto {
+
+  @ApiProperty({ example: "Reportko Project" })
+  @IsOptional()
+  @IsString()
+  name: string;
```
Comment (norikkozmelj) [inline src/model/dto/project.update.dto.ts:20]: if a field is optional then use ? when declaring a field.
Example:
name?: string;

File: src/module/project/project.controller.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  // .../project/id
+  @ApiBearerAuth()
+  @Roles(UserRoleKey.ADMIN)
+  @UseGuards(JwtAuthGuard, RoleGuard)
+  @Put(':id')
+  async updateProject(
+    @Param('id', ParseIntPipe) id: number,
+    @Body() projectCredentials: UpdateProjectDto
+  ): Promise<any> {
+    const project = await this.projectService.getOneById(id);
```
Comment (norikkozmelj) [inline src/module/project/project.controller.ts:68]: move this to service and send id in the function. All querying of the database should be done in service, not in controller.

File: src/module/project/project.controller.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  // Api response for updating project
+  @ApiCreatedResponse({
+    description: 'Project was succesfully updated'
+  })
+  @ApiUnauthorizedResponse({
+    description: 'Unauthorized'
+  })
+  // .../project/name/name
```
Comment (norikkozmelj) [inline src/module/project/project.controller.ts:79]: Not needed.

File: src/module/project/project.controller.ts [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+  @ApiUnauthorizedResponse({
+    description: 'Unauthorized'
+  })
+  // .../project/id
+  @ApiBearerAuth()
+  @Roles(UserRoleKey.ADMIN)
+  @UseGuards(JwtAuthGuard, RoleGuard)
+  @Delete(':id')
+  async deleteProject(@Param('id', ParseIntPipe) id: number): Promise<any> {
+    const project = await this.projectService.getOneById(id);
```
Comment (norikkozmelj) [inline src/module/project/project.controller.ts:106]: Same as with update.

File: src/module/project/project.controller.ts [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+
+  // Api response for deleting project
+  @ApiCreatedResponse({
+    description: 'Project was succesfully updated'
+  })
+  @ApiUnauthorizedResponse({
+    description: 'Unauthorized'
+  })
+  // .../project/name/name
+  @ApiBearerAuth()
```
Comment (norikkozmelj) [inline src/module/project/project.controller.ts:118]: Same here.

File: src/module/project/project.controller.ts [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+  async getOneById(@Param('id', ParseIntPipe) id: number): Promise<Project> {
+    return this.projectService.getOneById(Number(id));
+  }
+
+  // Api response for get project by project name
+  @ApiOkResponse({
+    description: 'Project with selected name is displayed'
+  })
+  // .../project/name/name
+  @ApiBearerAuth()
```
Comment (norikkozmelj) [inline src/module/project/project.controller.ts:158]: Not needed

File: src/module/project/project.service.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  ) {}
+
+  /**
+   * Function that creates user and adds it to database
+   * @param projectCredentials 
+   * @returns 
+   */
+  @Transactional()
+  async createProject(projectCredentials: CreateProjectDto): Promise<Project> {
+    projectCredentials.name = projectCredentials.name.toLowerCase();
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:24]: this can be done in entity with transform.

File: src/module/project/project.service.ts [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    const existingTeam = await this.getProjectByName(projectCredentials.name);
+
+    if (existingTeam) {
+      throw new ConflictException(ExceptionCodeName.INVALID_PROJECT_NAME);
+    }
+
+    project.name = projectCredentials.name;
+
+    const projectTeam = await this.teamService
+      .getTeamByTeamKey(projectCredentials.team);
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:38]: here you will use team id instead of key.

File: src/module/project/project.service.ts [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+    project.team = projectTeam;
+
+    console.log("tu");
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:41]: I am here :)

File: src/module/project/project.service.ts [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+   * @returns 
+   */
+  @Transactional()
+  async createProject(projectCredentials: CreateProjectDto): Promise<Project> {
+    projectCredentials.name = projectCredentials.name.toLowerCase();
+
+    const project = new Project();
+    project.active = projectCredentials.active;
+
+    const existingTeam = await this.getProjectByName(projectCredentials.name);
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:29]: existingProject

File: src/module/project/project.service.ts [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+  /**
+   * Function that updates project
+   * @param project 
+   * @param projectCredentials 
+   * @returns 
+   */
+  @Transactional()
+  async updateProject(
+    project: Project,
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:54]: here will be id

File: src/module/project/project.service.ts [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+      existingProject = await this.getProjectByName(projectCredentials.name);
+
+      if (existingProject) {
+        throw new ConflictException(ExceptionCodeName.INVALID_PROJECT_NAME);
+      }
+    }
+
+    var projectTeam;
+    if (projectCredentials.team) {
+      projectTeam = await this.teamService.getTeamByTeamKey(
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:70]: query by id and use let not var

File: src/module/project/project.service.ts [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+        projectCredentials.team
+      );
+      project.team = projectTeam;
+    }
+
+    // New team
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:76]: remove this

File: src/module/project/project.service.ts [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    const updateProject = await getRepository(Project)
+      .createQueryBuilder('project')
+      .update(Project)
+      .set(
+        { 
+          active: projectCredentials.active === undefined ? 
+            project.active :projectCredentials.active, 
+          name: existingProject ? project.name : projectCredentials.name
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:101]: Move this before query and then use just .update(project) without set.

File: src/module/project/project.service.ts [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+        }
+      )
+      .where('id = :id', { id: project.id })
+      .execute();
+
+    if (!updateProject) {
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:107]: Not needed.

File: src/module/project/project.service.ts [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+    return updateProject;
+  }
+
+  /**
+   * Function that deletes projects
+   * @param project 
+   * @returns 
+   */
+  @Transactional()
+  async deleteProject(project: Project): Promise<DeleteResult> {
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:120]: Here you will have id. First query for that project and if there is no project with that id throw error.

File: src/module/project/project.service.ts [hunk 24]
```diff
@@ -0,0 +1,189 @@
...
+   * Function that gets project by project name
+   * @param name 
+   * @returns 
+   */
+  @Transactional()
+  async getProjectByName(name: string): Promise<Project> {
+    return await getRepository(Project)
+    .createQueryBuilder('project')
+    .where('project.name = :name', { name: name })
+    .getOne();
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:146]: not neded

File: src/module/project/project.service.ts [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+      .getOne();
+
+    if (!project) {
+      throw new NotFoundException(ExceptionCodeName.INVALID_PROJECT_NAME);
+    }
+
+    return project;
+  }
+
+}
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:189]: not needed

**Review by norikkozmelj — CHANGES_REQUESTED** [review]

File: src/module/project/project.service.ts (hunk 24 above)
Comment (mokot) [reply src/module/project/project.service.ts:146]: leave this in service

**Review by mokot — COMMENTED** [review]
I have resolved problems

File: src/entity/project.entity.ts [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+  created_at: Date;
+
+  @Exclude()
+  @CreateDateColumn()
+  updated_at: Date;
+
+  @BeforeInsert()
+  nameToLowerCase() {
+    this.name = this.name.toLowerCase();
+  }
```
Comment (norikkozmelj) [inline src/entity/project.entity.ts:62]: Well done!

File: src/module/project/project.service.ts [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+  ) {}
+
+  /**
+   * Function that creates user and adds it to database
+   * @param projectCredentials 
+   * @returns 
+   */
+  @Transactional()
+  async createProject(projectCredentials: CreateProjectDto): Promise<Project> {
+    projectCredentials.name = projectCredentials.name.toLowerCase();
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:28]: This has to be removed as you covered this in the entity.

File: src/module/project/project.service.ts [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+
+    if (existingProject) {
+      throw new ConflictException(ExceptionCodeName.INVALID_PROJECT_NAME);
+    }
+
+    project.name = projectCredentials.name;
+
+    const projectTeam = await this.teamService
+      .getTeamByTeamId(projectCredentials.team_id);
+    project.team = projectTeam;
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:44]: What if the team with that id does not exist?

File: src/module/project/project.service.ts [hunk 29]
```diff
@@ -0,0 +1,152 @@
...
+   * @param projectCredentials 
+   * @returns 
+   */
+  @Transactional()
+  async updateProject(
+    id: number, 
+    projectCredentials: UpdateProjectDto
+  ): Promise<Project> {
+
+    const project = await this.getOneById(id);
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:61]: What if there is no project with that id?

File: src/module/project/project.service.ts [hunk 30]
```diff
@@ -0,0 +1,152 @@
...
+      existingProject = await this.getProjectByName(projectCredentials.name);
+
+      if (existingProject) {
+        throw new ConflictException(ExceptionCodeName.INVALID_PROJECT_NAME);
+      }
+    }
+
+    let projectTeam;
+    if (projectCredentials.team_id) {
+      projectTeam = await this.teamService.getTeamByTeamId(
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:74]: What if there is no team with that id?

File: src/module/project/project.service.ts [hunk 31]
```diff
@@ -0,0 +1,152 @@
...
+        projectCredentials.team_id
+      );
+    }
+
+    project.active = projectCredentials.active === undefined ? 
+      project.active : projectCredentials.active;
+    project.name = existingProject ? project.name : projectCredentials.name;
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:81]: you have to check if there is the name in the credentials not if the project already exists.

File: src/module/project/project.service.ts [hunk 32]
```diff
@@ -0,0 +1,152 @@
...
+    project.team = projectTeam ? projectTeam : project.team;
+
+    const updateProject = await getRepository(Project).save(project);
+
+    if (!updateProject) {
+      throw new NotFoundException(ExceptionCodeName.INVALID_CREDENTIALS);
+    }
+
+    return updateProject;
+  }
```
Comment (norikkozmelj) [inline src/module/project/project.service.ts:91]: If you cover all the possibilites, you do not need this check. Just return result of the await.

**Review by norikkozmelj — CHANGES_REQUESTED** [review]

File: src/module/project/project.service.ts (hunk 29 above)
Comment (mokot) [reply src/module/project/project.service.ts:61]: Same result as in getTeamBuTeamId (invalid_project_id)

File: src/module/project/project.service.ts (hunk 31 above)
Comment (mokot) [reply src/module/project/project.service.ts:81]: If there is name in project credentials and if that name is not equal to current name, then i check if there is project with that name in database. If project exists i throw conflict exception (invalid_project_name). If value of existing project is false/null/undefined i replace the name with project credentials name, else i keep the same name.

Comment (mokot) [reply src/module/project/project.service.ts:81]: Mb, if name is not in the credentials, function does not work ... I will fix the problem

File: src/entity/project.entity.ts [hunk 33]
```diff
@@ -0,0 +1,64 @@
...
+@Entity()
+export class Project {
+
+  @ApiProperty({ example: 1 })
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @ApiProperty({ example: "Reportko Project" })
+  @Column({ type: 'varchar', unique: true })
+  @IsString()
```
Comment (jakapurg) [inline src/entity/project.entity.ts:37]: No need for class validators in entity files. (Already handled in DTO's)

File: src/entity/project.entity.ts [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+
+  @ApiProperty()
+  @ManyToOne(() => Team)
+  @JoinColumn({ name: 'team_id' })
+  @Type(() => Team)
+  team: Team;
+
+  @ApiProperty({ example: true })
+  @Column({ type: 'boolean', default: true })
+  @IsBoolean()
```
Comment (jakapurg) [inline src/entity/project.entity.ts:48]: Same as above

File: src/model/dto/project.create.dto.ts [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+export class CreateProjectDto {
+
+  @ApiProperty({ example: "Reportko Project" })
+  @IsNotEmpty()
+  @IsString()
+  name: string;
+
+  @ApiProperty({ example: 1 })
+  @IsNotEmpty()
+  team_id: number;
```
Comment (jakapurg) [inline src/model/dto/project.create.dto.ts:23]: You can add decorator for positive integer here

File: src/model/dto/project.update.dto.ts [hunk 36]
```diff
@@ -0,0 +1,30 @@
...
+export class UpdateProjectDto {
+
+  @ApiProperty({ example: "Reportko Project" })
+  @IsOptional()
+  @IsString()
+  name?: string;
+
+  @ApiProperty({ example: 1 })
+  @IsOptional()
+  team_id?: number;
```
Comment (jakapurg) [inline src/model/dto/project.update.dto.ts:23]: Decorator for positive integer

File: src/module/project/project.controller.ts [hunk 37]
```diff
@@ -0,0 +1,113 @@
...
+  // Api response for get project by project id
+  @ApiOkResponse({
+    description: 'Project with selected id is displayed'
+  })
+  // .../project/id
+  @ApiBearerAuth()
+  @UseGuards(JwtAuthGuard)
+  @Get(':id')
+  async getOneById(@Param('id', ParseIntPipe) id: number): Promise<Project> {
+    return this.projectService.getOneById(Number(id));
```
Comment (jakapurg) [inline src/module/project/project.controller.ts:110]: Is Number() needed ?

File: src/module/project/project.service.ts [hunk 38]
```diff
@@ -0,0 +1,149 @@
...
+      .leftJoinAndSelect('project.team', 'team')
+      .where('project.id = :id', { id: id })
+      .getOne();
+
+    if (!project) {
+      throw new NotFoundException(ExceptionCodeName.INVALID_PROJECT_ID);
+    }
+
+    return project;
+  }
```
Comment (jakapurg) [inline src/module/project/project.service.ts:147]: Merge functions together. Hint: Use FindOneOptions

File: src/module/project/project.service.ts [hunk 39]
```diff
@@ -0,0 +1,149 @@
...
+   * Function that gets all projects
+   * @returns 
+   */
+  @Transactional()
+  async getAll(): Promise<Project[]> {
+    return getRepository(Project)
+      .createQueryBuilder('project')
+      .leftJoinAndSelect('project.team', 'team')
+      .getMany();
+  }
```
Comment (jakapurg) [inline src/module/project/project.service.ts:114]: Use FindManyOptions, for future queries

**Review by mokot — COMMENTED** [review]
Problems resolved.
I have not yet merged functions

**Review by mokot — COMMENTED** [review]
Comments Resolved

File: src/module/project/project.service.ts [hunk 40]
```diff
@@ -0,0 +1,150 @@
...
```
Comment (jakapurg) [inline src/module/project/project.service.ts:138]: Use FindManyOptions

File: src/module/project/project.controller.ts [hunk 41]
```diff
@@ -0,0 +1,126 @@
...
+  @ApiBearerAuth()
+  @UseGuards(JwtAuthGuard)
+  @Get(':id')
+  async getOneById(@Param('id', ParseIntPipe) id: number): Promise<Project> {
+    const project = await this.projectService.getOneBy(
+      {
+        where: { id: id },
+        relations: ['team']
+      }
+    );
```
Comment (jakapurg) [inline src/module/project/project.controller.ts:117]: NB: {id:id} is the same as {id}

File: src/module/project/project.service.ts [hunk 42]
```diff
@@ -0,0 +1,150 @@
...
+  @Transactional()
+  async updateProject(
+    id: number, 
+    projectCredentials: UpdateProjectDto
+  ): Promise<Project> {
+
+    const project = await this.getOneBy(
+      {
+        where: { id: id }
+      }
```
Comment (jakapurg) [inline src/module/project/project.service.ts:67]: NB: {id:id} is the same as {id}

File: src/module/project/project.service.ts [hunk 43]
```diff
@@ -0,0 +1,147 @@
...
+    return await getRepository(Project).delete(id);
+  }
+
+  /**
+   * Function that gets all projects
+   * @returns 
+   */
+  @Transactional()
+  async getAllBy(options: FindManyOptions): Promise<Project[]> {
+    return getRepository(Project).find(options);
```
Comment (mokot) [inline src/module/project/project.service.ts:134]: await

**Review by jakapurg — APPROVED** [review]

**Review by jakapurg — APPROVED** [review]

**Pull request merged**